/*global unexpected*/
// THIS FILE IS AUTOGENERATED! DO NOT CHANGE IT MANUALLY.
// It is built based on the examples in the documentation folder
// when the documentation site gets build by running "make site-build".
it.skipIf = function (condition) {
    (condition ? it.skip : it).apply(it, Array.prototype.slice.call(arguments, 1));
};

describe("documentation tests", function () {
    var isBrowser = typeof weknowhow !== 'undefined';
    var isPhantom = typeof mochaPhantomJS !== 'undefined';
    var expect;
    beforeEach(function () {
        expect = unexpected.clone();
        expect.output.preferredWidth = 80;

    });

    it("api/UnexpectedError.md contains correct examples", function () {
        var testPromises = [];
        expect.addAssertion("array", "to have item satisfying", function (expect, subject) {
          var args = Array.prototype.slice.call(arguments, 2);
          var promises = subject.map(function (item) {
            return expect.promise(function () {
              return expect.apply(expect, [item].concat(args));
            });
          });

          return expect.promise.settle(promises).then(function () {
            var failed = promises.every(function (promise) {
              return promise.isRejected();
            });

            if (failed) {
              expect.fail({
                diff: function (output, diff, inspect, equal) {
                  var result = {
                    inline: true,
                    diff: output
                  };
                  promises.forEach(function (promise, index) {
                    if (index > 0) { result.diff.nl(2); }
                    var error = promise.reason();
                    // the error is connected to the current scope
                    // but we are just interested in the nested error
                    error.errorMode = 'bubble';
                    result.diff.append(error.getErrorMessage(output));
                  });
                  return result;
                }
              });
            }
          });
        });

        try {
            expect(['foo', 'bar'], 'to have item satisfying', 'to equal', 'bar');
            expect(['foo', 'bar'], 'to have item satisfying', 'to equal', 'bAr');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(['foo', 'bar'], 'to have item satisfying', 'to equal', 'bar');").nl();
                output.code("expect(['foo', 'bar'], 'to have item satisfying', 'to equal', 'bAr');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 'foo', 'bar' ] to have item satisfying 'to equal', 'bAr'\n" +
                "\n" +
                "expected 'foo' to equal 'bAr'\n" +
                "\n" +
                "-foo\n" +
                "+bAr\n" +
                "\n" +
                "expected 'bar' to equal 'bAr'\n" +
                "\n" +
                "-bar\n" +
                "+bAr"
            );
        }

        try {
            expect.addAssertion('detailed to be', function (expect, subject, value) {
              expect.errorMode = 'bubble';
              expect.withError(function () {
                expect(subject, 'to be', value);
              }, function (err) {
                err.getParents().forEach(function (e) {
                  e.errorMode = 'nested';
                });
                expect.fail(err);
              });
            });

            expect('f00!', 'detailed to be', 'foo!');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.addAssertion('detailed to be', function (expect, subject, value) {").nl();
                output.code("  expect.errorMode = 'bubble';").nl();
                output.code("  expect.withError(function () {").nl();
                output.code("    expect(subject, 'to be', value);").nl();
                output.code("  }, function (err) {").nl();
                output.code("    err.getParents().forEach(function (e) {").nl();
                output.code("      e.errorMode = 'nested';").nl();
                output.code("    });").nl();
                output.code("    expect.fail(err);").nl();
                output.code("  });").nl();
                output.code("});").nl();
                output.code("").nl();
                output.code("expect('f00!', 'detailed to be', 'foo!');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'f00!' to be 'foo!'\n" +
                "  expected 'f00!' to equal 'foo!'\n" +
                "    Explicit failure\n" +
                "\n" +
                "    -f00!\n" +
                "    +foo!"
            );
        }

        try {
            expect.addAssertion('to be completely custom', function (expect, subject) {
              return expect.withError(function () {
                expect(subject, 'to satisfy', { custom: true });
              }, function (err) {
                var createDiff = err.getDiffMethod();
                expect.fail({
                  diff: function (output, diff, inspect, equal) {
                    output.text('~~~~~~~~~~~~~~').sp().success('custom').sp().text('~~~~~~~~~~~~~~').nl();
                    var result = createDiff(output, diff, inspect, equal);
                    return result;
                  }
                });
              });
            });

            expect({ custom: false }, 'to be completely custom');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.addAssertion('to be completely custom', function (expect, subject) {").nl();
                output.code("  return expect.withError(function () {").nl();
                output.code("    expect(subject, 'to satisfy', { custom: true });").nl();
                output.code("  }, function (err) {").nl();
                output.code("    var createDiff = err.getDiffMethod();").nl();
                output.code("    expect.fail({").nl();
                output.code("      diff: function (output, diff, inspect, equal) {").nl();
                output.code("        output.text('~~~~~~~~~~~~~~').sp().success('custom').sp().text('~~~~~~~~~~~~~~').nl();").nl();
                output.code("        var result = createDiff(output, diff, inspect, equal);").nl();
                output.code("        return result;").nl();
                output.code("      }").nl();
                output.code("    });").nl();
                output.code("  });").nl();
                output.code("});").nl();
                output.code("").nl();
                output.code("expect({ custom: false }, 'to be completely custom');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { custom: false } to be completely custom\n" +
                "\n" +
                "~~~~~~~~~~~~~~ custom ~~~~~~~~~~~~~~\n" +
                "{\n" +
                "  custom: false // should equal true\n" +
                "}"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("api/addAssertion.md contains correct examples", function () {
        var testPromises = [];
        var errorMode = 'default'; // use to control the error mode later in the example
        expect.addAssertion('array', '[not] to be (sorted|ordered)', function (expect, subject, cmp) {
          expect.errorMode = errorMode;
          expect(subject, '[not] to equal', [].concat(subject).sort(cmp));
        });

        expect([1,2,3], 'to be sorted');
        expect([1,2,3], 'to be ordered');
        expect([2,1,3], 'not to be sorted');
        expect([2,1,3], 'not to be ordered');
        expect([3,2,1], 'to be sorted', function (x, y) { return y - x; });

        try {
            expect.addAssertion('number', 'to be contained by', function (expect, subject, start, finish) {
                expect.subjectOutput = function (output) {
                    output.text('point ').jsNumber(subject);
                };
                expect.argsOutput = function (output) {
                    output.text('interval ').text('[').appendInspected(start).text(';').appendInspected(finish).text(']');
                };
                expect(subject >= start && subject <= finish, '[not] to be truthy');
            });

            expect(4, 'to be contained by', 8, 10);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.addAssertion('number', 'to be contained by', function (expect, subject, start, finish) {").nl();
                output.code("    expect.subjectOutput = function (output) {").nl();
                output.code("        output.text('point ').jsNumber(subject);").nl();
                output.code("    };").nl();
                output.code("    expect.argsOutput = function (output) {").nl();
                output.code("        output.text('interval ').text('[').appendInspected(start).text(';').appendInspected(finish).text(']');").nl();
                output.code("    };").nl();
                output.code("    expect(subject >= start && subject <= finish, '[not] to be truthy');").nl();
                output.code("});").nl();
                output.code("").nl();
                output.code("expect(4, 'to be contained by', 8, 10);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected point 4 to be contained by interval [8;10]"
            );
        }

        try {
            expect.addAssertion('number', 'to be similar to', function (expect, subject, value, epsilon) {
                if (typeof epsilon !== 'number') {
                    epsilon = 1e-9;
                }
                expect.argsOutput[2] = function (output) {
                    output.text('(epsilon: ').jsNumber(epsilon.toExponential()).text(')')
                }
                expect(Math.abs(subject - value), 'to be less than or equal to', epsilon);
            });

            expect(4, 'to be similar to', 4.0001);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.addAssertion('number', 'to be similar to', function (expect, subject, value, epsilon) {").nl();
                output.code("    if (typeof epsilon !== 'number') {").nl();
                output.code("        epsilon = 1e-9;").nl();
                output.code("    }").nl();
                output.code("    expect.argsOutput[2] = function (output) {").nl();
                output.code("        output.text('(epsilon: ').jsNumber(epsilon.toExponential()).text(')')").nl();
                output.code("    }").nl();
                output.code("    expect(Math.abs(subject - value), 'to be less than or equal to', epsilon);").nl();
                output.code("});").nl();
                output.code("").nl();
                output.code("expect(4, 'to be similar to', 4.0001);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 4 to be similar to 4.0001, (epsilon: 1e-9)"
            );
        }

        try {
            expect([ 1, 3, 2, 4 ], 'to be sorted');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([ 1, 3, 2, 4 ], 'to be sorted');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 3, 2, 4 ] to be sorted\n" +
                "\n" +
                "[\n" +
                "  1,\n" +
                "  3, // should equal 2\n" +
                "  2, // should equal 3\n" +
                "  4\n" +
                "]"
            );
        }

        try {
            errorMode = 'bubble';
            expect([ 1, 3, 2, 4 ], 'to be sorted');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("errorMode = 'bubble';").nl();
                output.code("expect([ 1, 3, 2, 4 ], 'to be sorted');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 3, 2, 4 ] to equal [ 1, 2, 3, 4 ]\n" +
                "\n" +
                "[\n" +
                "  1,\n" +
                "  3, // should equal 2\n" +
                "  2, // should equal 3\n" +
                "  4\n" +
                "]"
            );
        }

        try {
            errorMode = 'nested';
            expect([ 1, 3, 2, 4 ], 'to be sorted');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("errorMode = 'nested';").nl();
                output.code("expect([ 1, 3, 2, 4 ], 'to be sorted');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 3, 2, 4 ] to be sorted\n" +
                "  expected [ 1, 3, 2, 4 ] to equal [ 1, 2, 3, 4 ]\n" +
                "\n" +
                "  [\n" +
                "    1,\n" +
                "    3, // should equal 2\n" +
                "    2, // should equal 3\n" +
                "    4\n" +
                "  ]"
            );
        }

        try {
            errorMode = 'defaultOrNested';
            expect([ 1, 3, 2, 4 ], 'to be sorted');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("errorMode = 'defaultOrNested';").nl();
                output.code("expect([ 1, 3, 2, 4 ], 'to be sorted');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 3, 2, 4 ] to be sorted\n" +
                "\n" +
                "[\n" +
                "  1,\n" +
                "  3, // should equal 2\n" +
                "  2, // should equal 3\n" +
                "  4\n" +
                "]"
            );
        }

        try {
            errorMode = 'diff';
            expect([ 1, 3, 2, 4 ], 'to be sorted');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("errorMode = 'diff';").nl();
                output.code("expect([ 1, 3, 2, 4 ], 'to be sorted');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "[\n" +
                "  1,\n" +
                "  3, // should equal 2\n" +
                "  2, // should equal 3\n" +
                "  4\n" +
                "]"
            );
        }

        function Timelock(value, delay) {
          this.value = value;
          this.delay = delay;
        }

        Timelock.prototype.getValue = function (cb) {
          var that = this;
          setTimeout(function () {
            cb(that.value);
          }, this.delay);
        };

        expect.addType({
          name: 'Timelock',
          identify: function (value) {
            return value && value instanceof Timelock;
          },
          inspect: function (value, depth, output) {
            output.jsFunctionName('Timelock');
          }
        });

        expect.addAssertion('<Timelock> to satisfy <any|regexp|expect.it>', function (expect, subject, spec) {
          return expect.promise(function (run) {
            subject.getValue(run(function (value) {
              return expect(value, 'to satisfy', spec);
            }));
          });
        });

        testPromises.push(expect.promise(function () {
            return expect(new Timelock('Hello world!', 5), 'to satisfy', expect.it('not to match', /!/));
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect(new Timelock('Hello world!', 5), 'to satisfy', expect.it('not to match', /!/));").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Timelock to satisfy expect.it('not to match', /!/)\n" +
                "\n" +
                "expected 'Hello world!' not to match /!/\n" +
                "\n" +
                "Hello world!\n" +
                "           ^"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("api/addType.md contains correct examples", function () {
        var testPromises = [];
        function Person(name, age) {
            this.name = name;
            this.age = age;
        }

        expect.addType({
            name: 'Person',
            base: 'object',
            identify: function (value) {
                return value instanceof Person;
            }
        });

        try {
            expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected Person({ name: 'John Doe', age: 42 })\n" +
                "to equal Person({ name: 'Jane Doe', age: 24 })\n" +
                "\n" +
                "Person({\n" +
                "  name: 'John Doe', // should equal 'Jane Doe'\n" +
                "                    // -John Doe\n" +
                "                    // +Jane Doe\n" +
                "  age: 42 // should equal 24\n" +
                "})"
            );
        }

        expect = unexpected.clone();
        expect.output.preferredWidth = 80;

        expect.addType({
            name: 'Person',
            base: 'object',
            identify: function (value) {
                return value instanceof Person;
            },
            inspect: function (person, depth, output, inspect) {
               output.text('new Person(')
                     .append(inspect(person.name, depth))
                     .text(', ')
                     .append(inspect(person.age, depth))
                     .text(')');
            }
        });

        try {
            expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected new Person('John Doe', 42) to equal new Person('Jane Doe', 24)\n" +
                "\n" +
                "Person({\n" +
                "  name: 'John Doe', // should equal 'Jane Doe'\n" +
                "                    // -John Doe\n" +
                "                    // +Jane Doe\n" +
                "  age: 42 // should equal 24\n" +
                "})"
            );
        }

        expect = unexpected.clone();
        expect.output.preferredWidth = 80;

        expect.addType({
            name: 'Person',
            base: 'object',
            identify: function (value) {
                return value instanceof Person;
            },
            inspect: function (person, depth, output, inspect) {
               output.text('new Person(')
                     .append(inspect(person.name, depth))
                     .text(', ')
                     .append(inspect(person.age, depth))
                     .text(')');
            },
            equal: function (a, b, equal) {
                return a === b || equal(a.name, b.name);
            }
        });

        expect = unexpected.clone();
        expect.output.preferredWidth = 80;

        expect.addType({
            name: 'Person',
            base: 'object',
            identify: function (value) {
                return value instanceof Person;
            },
            inspect: function (person, depth, output, inspect) {
               output.text('new Person(')
                     .append(inspect(person.name, depth))
                     .text(', ')
                     .append(inspect(person.age, depth))
                     .text(')');
            },
            equal: function (a, b, equal) {
                return a === b || equal(a.name, b.name);
            },
            diff: function (actual, expected, output, diff, inspect) {
                return this.baseType.diff({name: actual.name}, {name: expected.name}, output);
            }
        });

        try {
            expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected new Person('John Doe', 42) to equal new Person('Jane Doe', 24)\n" +
                "\n" +
                "{\n" +
                "  name: 'John Doe' // should equal 'Jane Doe'\n" +
                "                   // -John Doe\n" +
                "                   // +Jane Doe\n" +
                "}"
            );
        }

        expect = unexpected.clone();
        expect.output.preferredWidth = 80;

        var inlineDiff = true; // used to change inlining in a later example

        expect.addType({
            name: 'Person',
            base: 'object',
            identify: function (value) {
                return value instanceof Person;
            },
            inspect: function (person, depth, output, inspect) {
               output.text('new Person(')
                     .append(inspect(person.name, depth))
                     .text(', ')
                     .append(inspect(person.age, depth))
                     .text(')');
            },
            equal: function (a, b, equal) {
                return a === b || equal(a.name, b.name);
            },
            diff: function (actual, expected, output, diff, inspect) {
                var nameDiff = diff(actual.name, expected.name);

                output.text('new Person(')
                      .nl()
                      .indentLines();

                if (nameDiff && nameDiff.inline) {
                    output.append(nameDiff.diff);
                } else {
                    output.i().append(inspect(actual.name)).text(',').sp()
                          .annotationBlock(function () {
                              this.error('should be ').append(inspect(expected.name));
                              if (nameDiff) {
                                  this.nl().append(nameDiff.diff);
                              }
                          })
                          .nl();
                }

                output.i().append(inspect(actual.age))
                      .outdentLines()
                      .nl()
                      .text(')');

                return {
                    inline: inlineDiff,
                    diff: output
                };
            }
        });

        try {
            expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new Person('John Doe', 42), 'to equal', new Person('Jane Doe', 24));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected new Person('John Doe', 42) to equal new Person('Jane Doe', 24)\n" +
                "\n" +
                "new Person(\n" +
                "  'John Doe', // should be 'Jane Doe'\n" +
                "              // -John Doe\n" +
                "              // +Jane Doe\n" +
                "  42\n" +
                ")"
            );
        }

        try {
            inlineDiff = true;
            expect(
              {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Janie Doe', 24)},
              'to equal',
              {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Jane Doe', 24)}
            );
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("inlineDiff = true;").nl();
                output.code("expect(").nl();
                output.code("  {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Janie Doe', 24)},").nl();
                output.code("  'to equal',").nl();
                output.code("  {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Jane Doe', 24)}").nl();
                output.code(");").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person('Janie Doe', 24)\n" +
                "}\n" +
                "to equal\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person('Jane Doe', 24)\n" +
                "}\n" +
                "\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person(\n" +
                "    'Janie Doe', // should be 'Jane Doe'\n" +
                "                 // -Janie Doe\n" +
                "                 // +Jane Doe\n" +
                "    24\n" +
                "  )\n" +
                "}"
            );
        }

        try {
            inlineDiff = false;
            expect(
              {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Janie Doe', 24)},
              'to equal',
              {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Jane Doe', 24)}
            );
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("inlineDiff = false;").nl();
                output.code("expect(").nl();
                output.code("  {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Janie Doe', 24)},").nl();
                output.code("  'to equal',").nl();
                output.code("  {'John Doe': new Person('John Doe', 42), 'Jane Doe': new Person('Jane Doe', 24)}").nl();
                output.code(");").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person('Janie Doe', 24)\n" +
                "}\n" +
                "to equal\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person('Jane Doe', 24)\n" +
                "}\n" +
                "\n" +
                "{\n" +
                "  'John Doe': new Person('John Doe', 42),\n" +
                "  'Jane Doe': new Person('Janie Doe', 24) // should equal new Person('Jane Doe', 24)\n" +
                "                                          // new Person(\n" +
                "                                          //   'Janie Doe', // should be 'Jane Doe'\n" +
                "                                          //                // -Janie Doe\n" +
                "                                          //                // +Jane Doe\n" +
                "                                          //   24\n" +
                "                                          // )\n" +
                "}"
            );
        }

        expect.addAssertion('Person', 'to be above legal age', function (expect, subject) {
            expect(subject.age, 'to be greater than or equal to', 18);
        });

        expect(new Person('Jane Doe', 24), 'to be above legal age');

        expect(new Person('Jane Doe', 24), 'to have keys', 'name', 'age');
        expect(new Person('Jane Doe', 24), 'to satisfy', {
            name: expect.it('to be a string').and('not to be empty'),
            age: expect.it('to be a number').and('not to be negative')
        });
        return expect.promise.all(testPromises);
    });

    it("api/clone.md contains correct examples", function () {
        var testPromises = [];
        var originalExpect = expect;

        expect = expect.clone().addAssertion('to be an integer', function (expect, subject) {
          expect(subject, 'to be a number');
          expect(Math.round(subject), 'to be', subject);
        });

        expect(42, 'to be an integer');

        try {
            originalExpect(42, 'to be an integer');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("originalExpect(42, 'to be an integer');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "Unknown assertion 'to be an integer', did you mean: 'to be infinite'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("api/expect.md contains correct examples", function () {
        var testPromises = [];
        try {
            expect(123, 'to equal', 456);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(123, 'to equal', 456);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 123 to equal 456"
            );
        }

        it('should call the callback', function () {
            return expect(setImmediate, 'to call the callback');
        });

        expect('abc', 'to be a string').and('to have length', 3);

        it('should do the right thing', function () {
            return expect(setImmediate, 'to be a function').and('to call the callback');
        });
        return expect.promise.all(testPromises);
    });

    it("api/fail.md contains correct examples", function () {
        var testPromises = [];
        try {
            expect.fail();
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.fail();").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "Explicit failure"
            );
        }

        try {
            expect.fail('Custom failure message');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.fail('Custom failure message');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "Custom failure message"
            );
        }

        try {
            expect.fail('{0} was expected to be {1}', 0, 'zero');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.fail('{0} was expected to be {1}', 0, 'zero');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "0 was expected to be 'zero'"
            );
        }

        try {
            var error = new Error('throw me');
            expect.fail(new Error(error));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("var error = new Error('throw me');").nl();
                output.code("expect.fail(new Error(error));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "Error: throw me"
            );
        }

        try {
            expect.fail(function (output) {
              'You have been a very bad boy!'.split(/ /).forEach(function (word, index) {
                if (index > 0) { output.sp(); }
                var style = index % 2 === 0 ? 'jsPrimitive' : 'jsString';
                output[style](word);
              });
            });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.fail(function (output) {").nl();
                output.code("  'You have been a very bad boy!'.split(/ /).forEach(function (word, index) {").nl();
                output.code("    if (index > 0) { output.sp(); }").nl();
                output.code("    var style = index % 2 === 0 ? 'jsPrimitive' : 'jsString';").nl();
                output.code("    output[style](word);").nl();
                output.code("  });").nl();
                output.code("});").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "You have been a very bad boy!"
            );
        }

        try {
            expect.fail({
              message: function (output) {
                'You have been a very bad boy!'.split(/ /).forEach(function (word, index) {
                  if (index > 0) { output.sp(); }
                  var style = index % 2 === 0 ? 'jsPrimitive' : 'jsString';
                  output[style](word);
                });
              },
              diff: function (output, diff, inspect, equal) {
                return diff('You have been a very bad boy!', 'You have been a very mad boy!');
              }
            });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect.fail({").nl();
                output.code("  message: function (output) {").nl();
                output.code("    'You have been a very bad boy!'.split(/ /).forEach(function (word, index) {").nl();
                output.code("      if (index > 0) { output.sp(); }").nl();
                output.code("      var style = index % 2 === 0 ? 'jsPrimitive' : 'jsString';").nl();
                output.code("      output[style](word);").nl();
                output.code("    });").nl();
                output.code("  },").nl();
                output.code("  diff: function (output, diff, inspect, equal) {").nl();
                output.code("    return diff('You have been a very bad boy!', 'You have been a very mad boy!');").nl();
                output.code("  }").nl();
                output.code("});").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "You have been a very bad boy!\n" +
                "\n" +
                "-You have been a very bad boy!\n" +
                "+You have been a very mad boy!"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("api/promise-all.md contains correct examples", function () {
        var testPromises = [];
        expect.addAssertion('to be a number after a short delay', function (expect, subject) {
          return expect.promise(function (run) {
            setTimeout(run(function () {
                expect(subject, 'to be a number');
            }), 1);
          });
        });

        testPromises.push(expect.promise(function () {
            return expect.promise.all({
              foo: [
                expect(42, 'to be a number after a short delay')
              ],
              bar: expect([0, 1, 2], 'to have items satisfying',
                                     expect.it('to be a number after a short delay')),

              baz: expect({ a: 1, b: 2 }, 'to have values satisfying',
                                          'to be a number after a short delay')
            });
        }));

        testPromises.push(expect.promise(function () {
            return expect.promise.all({
              foo: [
                expect(42, 'to be a number after a short delay')
              ],
              bar: expect([0, 1, 2], 'to have items satisfying',
                                     expect.it('to be a number after a short delay')),

              baz: expect({ a: '0', b: 1 }, 'to have values satisfying',
                                            'to be a number after a short delay')
            });
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect.promise.all({").nl();
                    output.code("  foo: [").nl();
                    output.code("    expect(42, 'to be a number after a short delay')").nl();
                    output.code("  ],").nl();
                    output.code("  bar: expect([0, 1, 2], 'to have items satisfying',").nl();
                    output.code("                         expect.it('to be a number after a short delay')),").nl();
                    output.code("").nl();
                    output.code("  baz: expect({ a: '0', b: 1 }, 'to have values satisfying',").nl();
                    output.code("                                'to be a number after a short delay')").nl();
                    output.code("});").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected { a: '0', b: 1 }\n" +
                "to have values satisfying 'to be a number after a short delay'\n" +
                "\n" +
                "{\n" +
                "  a: '0', // should be a number after a short delay\n" +
                "  b: 1\n" +
                "}"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("api/promise-any.md contains correct examples", function () {
        var testPromises = [];
        expect.addAssertion('to be a number after a short delay', function (expect, subject) {
          return expect.promise(function (run) {
            setTimeout(run(function () {
                expect(subject, 'to be a number');
            }), 0);
          });
        });

        testPromises.push(expect.promise(function () {
            return expect.promise.any({
              foo: [
                expect('42', 'to be a number after a short delay')
              ],
              bar: expect([0, 1, 2], 'to have items satisfying',
                                     expect.it('to be a number after a short delay')),

              baz: expect({ a: '1', b: 2 }, 'to have values satisfying',
                                            'to be a number after a short delay')
            });
        }));

        testPromises.push(expect.promise(function () {
            return expect.promise.any({
              foo: [
                expect('42', 'to be a number after a short delay')
              ],
              bar: expect([0, '1', 2], 'to have items satisfying',
                                       expect.it('to be a number after a short delay')),

              baz: expect({ a: '0', b: 1 }, 'to have values satisfying',
                                            'to be a number after a short delay')
            }).caught(function (aggregateError) {
              // Let's reformat the error a bit
              expect.fail(function (output) {
                output.error(aggregateError.message);
                var errors = [];
                for (var i = 0; i < aggregateError.length; i += 1) {
                  errors.push(aggregateError[i]);
                }

                errors.sort(function (a, b) { // Make the output stable
                  if (a.message < b.message) return -1;
                  if (a.message > b.message) return 1;
                  return 0;
                });

                output.indentLines();
                errors.forEach(function (e, i) {
                  output.nl().i().text(i + ': ').block(e.getErrorMessage({ output: output }));
                });
              });
            });
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect.promise.any({").nl();
                    output.code("  foo: [").nl();
                    output.code("    expect('42', 'to be a number after a short delay')").nl();
                    output.code("  ],").nl();
                    output.code("  bar: expect([0, '1', 2], 'to have items satisfying',").nl();
                    output.code("                           expect.it('to be a number after a short delay')),").nl();
                    output.code("").nl();
                    output.code("  baz: expect({ a: '0', b: 1 }, 'to have values satisfying',").nl();
                    output.code("                                'to be a number after a short delay')").nl();
                    output.code("}).caught(function (aggregateError) {").nl();
                    output.code("  // Let's reformat the error a bit").nl();
                    output.code("  expect.fail(function (output) {").nl();
                    output.code("    output.error(aggregateError.message);").nl();
                    output.code("    var errors = [];").nl();
                    output.code("    for (var i = 0; i < aggregateError.length; i += 1) {").nl();
                    output.code("      errors.push(aggregateError[i]);").nl();
                    output.code("    }").nl();
                    output.code("").nl();
                    output.code("    errors.sort(function (a, b) { // Make the output stable").nl();
                    output.code("      if (a.message < b.message) return -1;").nl();
                    output.code("      if (a.message > b.message) return 1;").nl();
                    output.code("      return 0;").nl();
                    output.code("    });").nl();
                    output.code("").nl();
                    output.code("    output.indentLines();").nl();
                    output.code("    errors.forEach(function (e, i) {").nl();
                    output.code("      output.nl().i().text(i + ': ').block(e.getErrorMessage({ output: output }));").nl();
                    output.code("    });").nl();
                    output.code("  });").nl();
                    output.code("});").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "aggregate error\n" +
                "  0: expected '42' to be a number after a short delay\n" +
                "  1: expected [ 0, '1', 2 ]\n" +
                "     to have items satisfying expect.it('to be a number after a short delay')\n" +
                "\n" +
                "     [\n" +
                "       0,\n" +
                "       '1', // should be a number after a short delay\n" +
                "       2\n" +
                "     ]\n" +
                "  2: expected { a: '0', b: 1 }\n" +
                "     to have values satisfying 'to be a number after a short delay'\n" +
                "\n" +
                "     {\n" +
                "       a: '0', // should be a number after a short delay\n" +
                "       b: 1\n" +
                "     }"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("api/promise-settle.md contains correct examples", function () {
        var testPromises = [];
        expect.addAssertion('to be a number after a short delay', function (expect, subject) {
          return expect.promise(function (run) {
            setTimeout(run(function () {
                expect(subject, 'to be a number');
            }), 1);
          });
        });

        testPromises.push(expect.promise(function () {
            var promises = {
              foo: expect('42', 'to be a number after a short delay'),
              bar: expect([0, 1, 2], 'to have items satisfying',
                                     expect.it('to be a number after a short delay')),
              baz: expect({ a: '1', b: 2 }, 'to have values satisfying',
                                            'to be a number after a short delay')
            };

            return expect.promise.all(promises).caught(function () {
              return expect.promise.settle(promises).then(function () {
                expect.fail(function (output) {
                  output.text('{').nl();
                  output.indentLines();
                  Object.keys(promises).forEach(function (key, index) {
                    output.i().jsKey(key).text(':').sp();
                    if (promises[key].isFulfilled()) {
                      output.success('✓');
                    } else {
                      output.error('⨯ ').block(promises[key].reason().getErrorMessage({ output: output }));
                    }
                    output.nl();
                  });
                  output.outdentLines();
                  output.text('}');
                });
              });
            });
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("var promises = {").nl();
                    output.code("  foo: expect('42', 'to be a number after a short delay'),").nl();
                    output.code("  bar: expect([0, 1, 2], 'to have items satisfying',").nl();
                    output.code("                         expect.it('to be a number after a short delay')),").nl();
                    output.code("  baz: expect({ a: '1', b: 2 }, 'to have values satisfying',").nl();
                    output.code("                                'to be a number after a short delay')").nl();
                    output.code("};").nl();
                    output.code("").nl();
                    output.code("return expect.promise.all(promises).caught(function () {").nl();
                    output.code("  return expect.promise.settle(promises).then(function () {").nl();
                    output.code("    expect.fail(function (output) {").nl();
                    output.code("      output.text('{').nl();").nl();
                    output.code("      output.indentLines();").nl();
                    output.code("      Object.keys(promises).forEach(function (key, index) {").nl();
                    output.code("        output.i().jsKey(key).text(':').sp();").nl();
                    output.code("        if (promises[key].isFulfilled()) {").nl();
                    output.code("          output.success('✓');").nl();
                    output.code("        } else {").nl();
                    output.code("          output.error('⨯ ').block(promises[key].reason().getErrorMessage({ output: output }));").nl();
                    output.code("        }").nl();
                    output.code("        output.nl();").nl();
                    output.code("      });").nl();
                    output.code("      output.outdentLines();").nl();
                    output.code("      output.text('}');").nl();
                    output.code("    });").nl();
                    output.code("  });").nl();
                    output.code("});").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "{\n" +
                "  foo: ⨯ expected '42' to be a number after a short delay\n" +
                "  bar: ✓\n" +
                "  baz: ⨯ expected { a: '1', b: 2 }\n" +
                "         to have values satisfying 'to be a number after a short delay'\n" +
                "\n" +
                "         {\n" +
                "           a: '1', // should be a number after a short delay\n" +
                "           b: 2\n" +
                "         }\n" +
                "}"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("api/use.md contains correct examples", function () {
        var testPromises = [];
        function IntegerInterval(from, to) {
          this.from = from;
          this.to = to;
        }

        expect.use({
          name: 'unexpected-integer-intervals',
          installInto: function (expect) {
              expect.addType({
                name: 'IntegerInterval',
                base: 'object',
                identify: function (value) {
                  return value && value instanceof IntegerInterval;
                },
                inspect: function (value, depth, output) {
                  output.text('[').jsNumber(value.from).text(',').jsNumber(value.to).text(']');
                }
              });

             expect.addAssertion('[not] to contain', function (expect, subject, value) {
               expect(value, '[not] to be within', subject.from, subject.to);
             });
          }
        });

        expect(new IntegerInterval(7, 13), 'to contain', 9);

        try {
            expect(new IntegerInterval(7, 13), 'to contain', 27);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new IntegerInterval(7, 13), 'to contain', 27);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [7,13] to contain 27"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("api/withError.md contains correct examples", function () {
        var testPromises = [];
        try {
            function Person(options) {
              this.name = options.name;
              this.gender = options.gender;
            }

            Person.prototype.genderSign = function () {
              switch (this.gender) {
              case 'female': return '♀';
              case 'male': return '♂';
              default: return '⚧';
              }
            };

            expect.addAssertion('to have same gender as', function (expect, subject, value) {
              expect.withError(function () {
                expect(subject.gender, 'to be', value.gender);
              }, function (e) {
                expect.fail({
                  diff: function (output) {
                    return {
                      inline: false,
                      diff: output.bold(subject.genderSign()).text(' ≠ ').bold(value.genderSign())
                    };
                  }
                });
              });
            });

            expect(new Person({ name: 'John Doe', gender: 'male' }),
                   'to have same gender as',
                   new Person({ name: 'Jane Doe', gender: 'female' }));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("function Person(options) {").nl();
                output.code("  this.name = options.name;").nl();
                output.code("  this.gender = options.gender;").nl();
                output.code("}").nl();
                output.code("").nl();
                output.code("Person.prototype.genderSign = function () {").nl();
                output.code("  switch (this.gender) {").nl();
                output.code("  case 'female': return '♀';").nl();
                output.code("  case 'male': return '♂';").nl();
                output.code("  default: return '⚧';").nl();
                output.code("  }").nl();
                output.code("};").nl();
                output.code("").nl();
                output.code("expect.addAssertion('to have same gender as', function (expect, subject, value) {").nl();
                output.code("  expect.withError(function () {").nl();
                output.code("    expect(subject.gender, 'to be', value.gender);").nl();
                output.code("  }, function (e) {").nl();
                output.code("    expect.fail({").nl();
                output.code("      diff: function (output) {").nl();
                output.code("        return {").nl();
                output.code("          inline: false,").nl();
                output.code("          diff: output.bold(subject.genderSign()).text(' ≠ ').bold(value.genderSign())").nl();
                output.code("        };").nl();
                output.code("      }").nl();
                output.code("    });").nl();
                output.code("  });").nl();
                output.code("});").nl();
                output.code("").nl();
                output.code("expect(new Person({ name: 'John Doe', gender: 'male' }),").nl();
                output.code("       'to have same gender as',").nl();
                output.code("       new Person({ name: 'Jane Doe', gender: 'female' }));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected Person({ name: 'John Doe', gender: 'male' })\n" +
                "to have same gender as Person({ name: 'Jane Doe', gender: 'female' })\n" +
                "\n" +
                "♂ ≠ ♀"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/Buffer/when-decoded-as.md contains correct examples", function () {
        var testPromises = [];
        if (!isBrowser) {
            expect(new Buffer([0xe2, 0x98, 0xba]), 'when decoded as', 'utf-8', 'to equal', '☺');
        }

        if (!isBrowser) {
            try {
                expect(new Buffer([0xe2, 0x98, 0xba]), 'when decoded as', 'utf-8', 'to equal', 'happy face');
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(new Buffer([0xe2, 0x98, 0xba]), 'when decoded as', 'utf-8', 'to equal', 'happy face');").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected Buffer([0xE2, 0x98, 0xBA]) when decoded as 'utf-8' to equal 'happy face'\n" +
                    "\n" +
                    "-☺\n" +
                    "+happy face"
                );
            }
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/Error/to-have-message.md contains correct examples", function () {
        var testPromises = [];
        expect(new Error('foobar'), 'to have message', 'foobar');

        try {
            expect(new Error('foobar'), 'to have message', 'barfoo');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(new Error('foobar'), 'to have message', 'barfoo');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected Error('foobar') to have message 'barfoo'\n" +
                "  expected 'foobar' to equal 'barfoo'\n" +
                "\n" +
                "  -foobar\n" +
                "  +barfoo"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/Promise/to-be-fulfilled.md contains correct examples", function () {
        var testPromises = [];
        testPromises.push(expect.promise(function () {
            var promiseThatWillBeFulfilled = expect.promise(function (resolve, reject) {
                setTimeout(resolve, 1);
            });

            return expect(promiseThatWillBeFulfilled, 'to be fulfilled');
        }));

        testPromises.push(expect.promise(function () {
            var rejectedPromise = expect.promise(function (resolve, reject) {
                setTimeout(function () {
                    reject(new Error('argh'));
                }, 1);
            });

            return expect(rejectedPromise, 'to be fulfilled');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("var rejectedPromise = expect.promise(function (resolve, reject) {").nl();
                    output.code("    setTimeout(function () {").nl();
                    output.code("        reject(new Error('argh'));").nl();
                    output.code("    }, 1);").nl();
                    output.code("});").nl();
                    output.code("").nl();
                    output.code("return expect(rejectedPromise, 'to be fulfilled');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (rejected) => Error('argh') to be fulfilled\n" +
                "  Promise (rejected) => Error('argh') unexpectedly rejected with Error('argh')"
            );
        }));


        var promiseThatWillBeFulfilledWithAValue = expect.promise(function (resolve, reject) {
            setTimeout(function () {
                resolve('abc');
            }, 1);
        });

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeFulfilledWithAValue, 'to be fulfilled with', 'abc');
        }));

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeFulfilledWithAValue, 'to be fulfilled with', /b/);
        }));

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeFulfilledWithAValue, 'to be fulfilled with', 'def');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect(promiseThatWillBeFulfilledWithAValue, 'to be fulfilled with', 'def');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (fulfilled) => 'abc' to be fulfilled with 'def'\n" +
                "  expected 'abc' to equal 'def'\n" +
                "\n" +
                "  -abc\n" +
                "  +def"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("assertions/Promise/to-be-rejected.md contains correct examples", function () {
        var testPromises = [];
        testPromises.push(expect.promise(function () {
            var promiseThatWillBeRejected = expect.promise(function (resolve, reject) {
                setTimeout(reject, 1);
            });

            return expect(promiseThatWillBeRejected, 'to be rejected');
        }));

        testPromises.push(expect.promise(function () {
            var fulfilledPromise = expect.promise(function (resolve, reject) {
                setTimeout(resolve, 1);
            });

            return expect(fulfilledPromise, 'to be rejected');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("var fulfilledPromise = expect.promise(function (resolve, reject) {").nl();
                    output.code("    setTimeout(resolve, 1);").nl();
                    output.code("});").nl();
                    output.code("").nl();
                    output.code("return expect(fulfilledPromise, 'to be rejected');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (fulfilled) to be rejected\n" +
                "  Promise (fulfilled) unexpectedly fulfilled"
            );
        }));


        var promiseThatWillBeRejectedWithAReason = expect.promise(function (resolve, reject) {
            setTimeout(function () {
                reject(new Error('Oh dear'));
            }, 1);
        });

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeRejectedWithAReason, 'to be rejected with', new Error('Oh dear'));
        }));

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeRejectedWithAReason, 'to be rejected with', /dear/);
        }));

        testPromises.push(expect.promise(function () {
            return expect(promiseThatWillBeRejectedWithAReason, 'to be rejected with', new Error('bugger'));
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect(promiseThatWillBeRejectedWithAReason, 'to be rejected with', new Error('bugger'));").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (rejected) => Error('Oh dear') to be rejected with Error('bugger')\n" +
                "  expected Error('Oh dear') to satisfy Error('bugger')\n" +
                "\n" +
                "  Error({\n" +
                "    message: 'Oh dear' // should equal 'bugger'\n" +
                "                       // -Oh dear\n" +
                "                       // +bugger\n" +
                "  })"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("assertions/Promise/when-fulfilled.md contains correct examples", function () {
        var testPromises = [];
        var fulfilledPromise = expect.promise(function (resolve, reject) {
            setTimeout(function () {
                resolve(123);
            });
        });

        testPromises.push(expect.promise(function () {
            return expect(fulfilledPromise, 'when fulfilled', 'to equal', 123);
        }));

        testPromises.push(expect.promise(function () {
            return expect(fulfilledPromise, 'when fulfilled', expect.it('to be greater than', 100));
        }));

        testPromises.push(expect.promise(function () {
            var rejectedPromise = expect.promise(function (resolve, reject) {
                setTimeout(function () {
                    reject(new Error('argh'));
                });
            });

            return expect(rejectedPromise, 'when fulfilled', 'to equal', 123);
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("var rejectedPromise = expect.promise(function (resolve, reject) {").nl();
                    output.code("    setTimeout(function () {").nl();
                    output.code("        reject(new Error('argh'));").nl();
                    output.code("    });").nl();
                    output.code("});").nl();
                    output.code("").nl();
                    output.code("return expect(rejectedPromise, 'when fulfilled', 'to equal', 123);").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (rejected) => Error('argh') when fulfilled to equal 123\n" +
                "  Promise (rejected) => Error('argh') unexpectedly rejected with Error('argh')"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("assertions/Promise/when-rejected.md contains correct examples", function () {
        var testPromises = [];
        var rejectedPromise = expect.promise(function (resolve, reject) {
            setTimeout(function () {
                reject(new Error('argh'));
            });
        });

        testPromises.push(expect.promise(function () {
            return expect(rejectedPromise, 'when rejected', 'to equal', new Error('argh'));
        }));

        testPromises.push(expect.promise(function () {
            return expect(rejectedPromise, 'when rejected', expect.it('to have message', 'argh'));
        }));

        testPromises.push(expect.promise(function () {
            var fulfilledPromise = expect.promise(function (resolve, reject) {
                setTimeout(function () {
                    resolve(123);
                });
            });

            return expect(fulfilledPromise, 'when rejected', 'to have message', 'argh');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("var fulfilledPromise = expect.promise(function (resolve, reject) {").nl();
                    output.code("    setTimeout(function () {").nl();
                    output.code("        resolve(123);").nl();
                    output.code("    });").nl();
                    output.code("});").nl();
                    output.code("").nl();
                    output.code("return expect(fulfilledPromise, 'when rejected', 'to have message', 'argh');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected Promise (fulfilled) => 123 when rejected to have message 'argh'\n" +
                "  Promise (fulfilled) => 123 unexpectedly fulfilled with 123"
            );
        }));

        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-a.md contains correct examples", function () {
        var testPromises = [];
        expect(true, 'to be a', 'boolean');
        expect(5, 'to be a', 'number');
        expect('abc', 'to be a', 'string');
        expect(expect, 'to be a', 'function');
        expect({foo: 123}, 'to be an', 'object');
        expect([123], 'to be an', 'array');
        expect(/regex/, 'to be a', 'regexp');
        expect(/regex/, 'to be a', 'regex');
        expect(/regex/, 'to be a', 'regular expression');
        expect(new Error(), 'to be an', 'Error');

        expect(expect, 'to be a', 'function');
        expect(expect, 'to be an', 'object');

        expect(true, 'to be a boolean');
        expect(5, 'to be a number');
        expect('abc', 'to be a string');
        expect(expect, 'to be a function');
        expect({foo: 123}, 'to be an object');
        expect([123], 'to be an array');
        expect(/regex/, 'to be a regexp');
        expect(/regex/, 'to be a regex');
        expect(/regex/, 'to be a regular expression');

        function Person(name) {
            this.name = name;
        }
        expect(new Person('John Doe'), 'to be a', Person);
        expect(new Person('John Doe'), 'to be an', Object);

        try {
            expect({ 0: 'foo', 1: 'bar', 2: 'baz' }, 'to be an array');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ 0: 'foo', 1: 'bar', 2: 'baz' }, 'to be an array');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { 0: 'foo', 1: 'bar', 2: 'baz' } to be an array"
            );
        }

        expect(true, 'not to be an object');
        expect('5', 'not to be a', 'number');
        expect('abc', 'not to be an', Object);

        try {
            expect(function () { return 'wat'; }, 'not to be an', Object);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(function () { return 'wat'; }, 'not to be an', Object);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected function () { return 'wat'; } not to be an Object"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-defined.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello world!', 'to be defined');
        expect({ foo: { bar: 'baz' } }, 'to be defined');

        try {
            expect(undefined, 'to be defined');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(undefined, 'to be defined');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected undefined to be defined"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-falsy.md contains correct examples", function () {
        var testPromises = [];
        expect(0, 'to be falsy');
        expect(false, 'to be falsy');
        expect('', 'to be falsy');
        expect(undefined, 'to be falsy');
        expect(null, 'to be falsy');

        try {
            expect({}, 'to be falsy');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({}, 'to be falsy');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected {} to be falsy"
            );
        }

        expect(1, 'not to be falsy');
        expect(true, 'not to be falsy');
        expect({}, 'not to be falsy');
        expect('foo', 'not to be falsy');
        expect(/foo/, 'not to be falsy');

        try {
            expect('', 'not to be falsy');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('', 'not to be falsy');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '' not to be falsy"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-null.md contains correct examples", function () {
        var testPromises = [];
        expect(null, 'to be null');

        try {
            expect({ foo: { bar: 'baz' } }, 'to be null');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ foo: { bar: 'baz' } }, 'to be null');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { foo: { bar: 'baz' } } to be null"
            );
        }

        expect({ foo: { bar: 'baz' } }, 'not to be null');
        expect('Hello world!', 'not to be null');

        try {
            expect(null, 'not to be null');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(null, 'not to be null');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected null not to be null"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-ok.md contains correct examples", function () {
        var testPromises = [];
        expect(1, 'to be ok');
        expect(true, 'to be ok');
        expect({}, 'to be ok');
        expect('foo', 'to be ok');
        expect(/foo/, 'to be ok');

        try {
            expect('', 'to be ok');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('', 'to be ok');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '' to be ok"
            );
        }

        expect(0, 'not to be ok');
        expect(false, 'not to be ok');
        expect('', 'not to be ok');
        expect(undefined, 'not to be ok');
        expect(null, 'not to be ok');

        try {
            expect({}, 'not to be ok');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({}, 'not to be ok');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected {} not to be ok"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-truthy.md contains correct examples", function () {
        var testPromises = [];
        expect(1, 'to be truthy');
        expect(true, 'to be truthy');
        expect({}, 'to be truthy');
        expect('foo', 'to be truthy');
        expect(/foo/, 'to be truthy');

        try {
            expect('', 'to be truthy');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('', 'to be truthy');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '' to be truthy"
            );
        }

        expect(0, 'not to be truthy');
        expect(false, 'not to be truthy');
        expect('', 'not to be truthy');
        expect(undefined, 'not to be truthy');
        expect(null, 'not to be truthy');

        try {
            expect({}, 'not to be truthy');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({}, 'not to be truthy');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected {} not to be truthy"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be-undefined.md contains correct examples", function () {
        var testPromises = [];
        expect(undefined, 'to be undefined');

        try {
            expect('Hello world', 'to be undefined');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world', 'to be undefined');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world' to be undefined"
            );
        }

        expect('Hello world!', 'not to be undefined');
        expect({ foo: { bar: 'baz' } }, 'not to be undefined');

        try {
            expect(undefined, 'not to be undefined');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(undefined, 'not to be undefined');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected undefined not to be undefined"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-be.md contains correct examples", function () {
        var testPromises = [];
        var obj = {};
        expect(obj, 'to be', obj);
        expect(1, 'to be', 1);
        expect(null, 'to be', null);
        expect(undefined, 'to be', obj.foo);
        expect(true, 'to be', !false);

        expect(NaN, 'to be', NaN);
        expect(-0, 'not to be', 0);

        try {
            expect('1', 'to be', 1);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('1', 'to be', 1);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '1' to be 1"
            );
        }

        expect({}, 'not to be', {});
        expect(1, 'not to be', true);
        expect('1', 'not to be', 1);
        expect(null, 'not to be', undefined);
        expect(0, 'not to be', 'null');
        expect(undefined, 'not to be', 'null');
        expect(false, 'not to be', 'true');
        expect(true, 'not to be', 'false');

        try {
            expect(1, 'not to be', 1);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1, 'not to be', 1);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1 not to be 1"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-equal.md contains correct examples", function () {
        var testPromises = [];
        expect({ a: 'b' }, 'to equal', { a: 'b' });
        var now = new Date();
        expect(now, 'to equal', now);
        expect(now, 'to equal', new Date(now.getTime()));
        expect({ now: now }, 'to equal', { now: now });

        try {
            expect({ text: 'foo!' }, 'to equal', { text: 'f00!' });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ text: 'foo!' }, 'to equal', { text: 'f00!' });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { text: 'foo!' } to equal { text: 'f00!' }\n" +
                "\n" +
                "{\n" +
                "  text: 'foo!' // should equal 'f00!'\n" +
                "               // -foo!\n" +
                "               // +f00!\n" +
                "}"
            );
        }

        try {
            expect({ one: 1, two: 2, four: 4, five: 5 }, 'to equal', { one: 1, two: 2, three: 3, four: 4 });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ one: 1, two: 2, four: 4, five: 5 }, 'to equal', { one: 1, two: 2, three: 3, four: 4 });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { one: 1, two: 2, four: 4, five: 5 }\n" +
                "to equal { one: 1, two: 2, three: 3, four: 4 }\n" +
                "\n" +
                "{\n" +
                "  one: 1,\n" +
                "  two: 2,\n" +
                "  four: 4,\n" +
                "  five: 5, // should be removed\n" +
                "  three: undefined // should equal 3\n" +
                "}"
            );
        }

        try {
            expect([ 0, 1, 2, 4, 5], 'to equal', [ 1, 2, 3, 4]);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([ 0, 1, 2, 4, 5], 'to equal', [ 1, 2, 3, 4]);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 0, 1, 2, 4, 5 ] to equal [ 1, 2, 3, 4 ]\n" +
                "\n" +
                "[\n" +
                "  0, // should be removed\n" +
                "  1,\n" +
                "  2,\n" +
                "  // missing 3\n" +
                "  4,\n" +
                "  5 // should be removed\n" +
                "]"
            );
        }

        if (!isBrowser) {
            try {
                expect(
                    new Buffer('\x00\x01\x02Here is the thing I was talking about', 'utf-8'),
                    'to equal',
                    new Buffer('\x00\x01\x02Here is the thing I was quuxing about', 'utf-8')
                );
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(").nl();
                    output.code("    new Buffer('\\x00\\x01\\x02Here is the thing I was talking about', 'utf-8'),").nl();
                    output.code("    'to equal',").nl();
                    output.code("    new Buffer('\\x00\\x01\\x02Here is the thing I was quuxing about', 'utf-8')").nl();
                    output.code(");").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected Buffer([0x00, 0x01, 0x02, 0x48, 0x65, 0x72, 0x65, 0x20, 0x69, 0x73, 0x20, 0x74, 0x68, 0x65, 0x20, 0x74 /* 24 more */ ])\n" +
                    "to equal Buffer([0x00, 0x01, 0x02, 0x48, 0x65, 0x72, 0x65, 0x20, 0x69, 0x73, 0x20, 0x74, 0x68, 0x65, 0x20, 0x74 /* 24 more */ ])\n" +
                    "\n" +
                    " 00 01 02 48 65 72 65 20 69 73 20 74 68 65 20 74  │...Here is the t│\n" +
                    "-68 69 6E 67 20 49 20 77 61 73 20 74 61 6C 6B 69  │hing I was talki│\n" +
                    "+68 69 6E 67 20 49 20 77 61 73 20 71 75 75 78 69  │hing I was quuxi│\n" +
                    " 6E 67 20 61 62 6F 75 74                          │ng about│"
                );
            }
        }

        expect(1, 'not to equal', '1');
        expect({ one: 1 }, 'not to equal', { one: '1' });
        expect(null, 'not to equal', '1');
        var now = new Date();
        var later = new Date(now.getTime() + 42);
        expect(now, 'not to equal', later);
        expect({ time: now }, 'not to equal', { time: later });

        try {
            expect({ a: { b: 'd'} }, 'not to equal', { a: { b: 'd'} });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: { b: 'd'} }, 'not to equal', { a: { b: 'd'} });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: { b: 'd' } } not to equal { a: { b: 'd' } }"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/to-satisfy.md contains correct examples", function () {
        var testPromises = [];
        expect({ hey: { there: true } }, 'to satisfy', { hey: {} });

        expect({ hey: { there: true } }, 'to exhaustively satisfy', { hey: { there: true } });

        expect({ bar: 'quux', baz: true }, 'to satisfy', { bar: /QU*X/i });

        expect({foo: 123, bar: 'bar', baz: 'bogus', qux: 42, quux: 'wat'}, 'to satisfy', {
            foo: expect.it('to be a number').and('to be greater than', 10),
            baz: expect.it('not to match', /^boh/),
            qux: expect.it('to be a string')
                          .and('not to be empty')
                       .or('to be a number')
                          .and('to be positive'),
            quux: function (value) {
              expect(value, 'to be a string');
            }
        });

        try {
            expect({foo: 9, bar: 'bar', baz: 'bogus', qux: 42, quux: 'wat'}, 'to satisfy', {
                foo: expect.it('to be a number').and('to be greater than', 10),
                baz: expect.it('not to match', /^bog/),
                qux: expect.it('to be a string')
                              .and('not to be empty')
                           .or('to be a number')
                              .and('to be positive'),
                quux: function (value) {
                  expect(value, 'to be a number');
                }
            });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({foo: 9, bar: 'bar', baz: 'bogus', qux: 42, quux: 'wat'}, 'to satisfy', {").nl();
                output.code("    foo: expect.it('to be a number').and('to be greater than', 10),").nl();
                output.code("    baz: expect.it('not to match', /^bog/),").nl();
                output.code("    qux: expect.it('to be a string')").nl();
                output.code("                  .and('not to be empty')").nl();
                output.code("               .or('to be a number')").nl();
                output.code("                  .and('to be positive'),").nl();
                output.code("    quux: function (value) {").nl();
                output.code("      expect(value, 'to be a number');").nl();
                output.code("    }").nl();
                output.code("});").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { foo: 9, bar: 'bar', baz: 'bogus', qux: 42, quux: 'wat' } to satisfy\n" +
                "{\n" +
                "  foo: expect.it('to be a number')\n" +
                "               .and('to be greater than', 10),\n" +
                "  baz: expect.it('not to match', /^bog/),\n" +
                "  qux: expect.it('to be a string')\n" +
                "               .and('not to be empty')\n" +
                "             .or('to be a number')\n" +
                "               .and('to be positive'),\n" +
                "  quux: function (value) {\n" +
                "    expect(value, 'to be a number');\n" +
                "  }\n" +
                "}\n" +
                "\n" +
                "{\n" +
                "  foo: 9, // ✓ should be a number and\n" +
                "          // ⨯ should be greater than 10\n" +
                "  bar: 'bar',\n" +
                "  baz: 'bogus', // should not match /^bog/\n" +
                "                //\n" +
                "                // bogus\n" +
                "                // ^^^\n" +
                "  qux: 42,\n" +
                "  quux: 'wat' // should be a number\n" +
                "}"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/any/when-passed-as-parameter-to.md contains correct examples", function () {
        var testPromises = [];
        function increment(n) {
            return n + 1;
        }

        expect(1, 'when passed as parameter to', increment, 'to equal', 2);

        try {
            expect(1, 'when passed as parameter to', increment, 'to equal', 3);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1, 'when passed as parameter to', increment, 'to equal', 3);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1\n" +
                "when passed as parameter to function increment(n) { return n + 1; } to equal 3\n" +
                "  expected 2 to equal 3"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/to-be-empty.md contains correct examples", function () {
        var testPromises = [];
        expect([], 'to be empty');

        try {
            expect([1,2,3], 'to be empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([1,2,3], 'to be empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 2, 3 ] to be empty"
            );
        }

        expect([1,2,3], 'not to be empty');

        try {
            expect([], 'not to be empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([], 'not to be empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [] not to be empty"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/to-be-non-empty.md contains correct examples", function () {
        var testPromises = [];
        expect([1, 2, 3], 'to be non-empty');

        try {
            expect([], 'to be non-empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([], 'to be non-empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [] to be non-empty"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/to-contain.md contains correct examples", function () {
        var testPromises = [];
        expect([0, 1, 2], 'to contain', 1);
        expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'to contain', { name: 'Jane Doe' });
        expect([0, 1, 2], 'to contain', 0, 2);

        try {
            expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'to contain', { name: 'Jonnie Doe' });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'to contain', { name: 'Jonnie Doe' });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ { name: 'John Doe' }, { name: 'Jane Doe' } ]\n" +
                "to contain { name: 'Jonnie Doe' }"
            );
        }

        expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'not to contain', { name: 'Jonnie Doe' });

        try {
            expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'not to contain', { name: 'Jane Doe' });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([ { name: 'John Doe' }, { name: 'Jane Doe' } ], 'not to contain', { name: 'Jane Doe' });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ { name: 'John Doe' }, { name: 'Jane Doe' } ]\n" +
                "not to contain { name: 'Jane Doe' }\n" +
                "\n" +
                "[\n" +
                "  { name: 'John Doe' },\n" +
                "  { name: 'Jane Doe' } // should be removed\n" +
                "]"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/to-have-items-satisfying.md contains correct examples", function () {
        var testPromises = [];
        expect([0, 1, 2, 3, 4], 'to have items satisfying', function (item, index) {
            expect(item, 'to be a number');
        });

        expect([0, 1, 2, 3, 4], 'to have items satisfying', 'to be a number');

        expect([[1], [2]], 'to have items satisfying',
               'to have items satisfying', 'to be a number');

        expect([1, 2, 3, 4], 'to have items satisfying',
          expect.it('to be a number').and('to be positive'));

        try {
            expect([ [0, 1, 2], [4, '5', '6'], [7, '8', 9] ],
              'to have items satisfying',
              'to have items satisfying',
              'to be a number');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([ [0, 1, 2], [4, '5', '6'], [7, '8', 9] ],").nl();
                output.code("  'to have items satisfying',").nl();
                output.code("  'to have items satisfying',").nl();
                output.code("  'to be a number');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected array to have items satisfying 'to have items satisfying', 'to be a number'\n" +
                "\n" +
                "[\n" +
                "  [ 0, 1, 2 ],\n" +
                "  [\n" +
                "    4,\n" +
                "    '5', // should be a number\n" +
                "    '6' // should be a number\n" +
                "  ],\n" +
                "  [\n" +
                "    7,\n" +
                "    '8', // should be a number\n" +
                "    9\n" +
                "  ]\n" +
                "]"
            );
        }

        try {
            expect([0, 1, 2, 3, 4], 'to have items satisfying',
              expect.it('to be a number').and('to be positive'));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([0, 1, 2, 3, 4], 'to have items satisfying',").nl();
                output.code("  expect.it('to be a number').and('to be positive'));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 0, 1, 2, 3, 4 ] to have items satisfying\n" +
                "expect.it('to be a number')\n" +
                "        .and('to be positive')\n" +
                "\n" +
                "[\n" +
                "  0, // ✓ should be a number and\n" +
                "     // ⨯ should be positive\n" +
                "  1,\n" +
                "  2,\n" +
                "  3,\n" +
                "  4\n" +
                "]"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/to-have-length.md contains correct examples", function () {
        var testPromises = [];
        expect([1,2,3], 'to have length', 3);

        try {
            expect([1,2,3], 'to have length', 4);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([1,2,3], 'to have length', 4);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 2, 3 ] to have length 4\n" +
                "  expected 3 to be 4"
            );
        }

        expect([1,2,3], 'not to have length', 4);

        try {
            expect([1,2,3], 'not to have length', 3);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([1,2,3], 'not to have length', 3);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 2, 3 ] not to have length 3"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/array-like/when-passed-as-parameters-to.md contains correct examples", function () {
        var testPromises = [];
        function add(a, b) {
            return a + b;
        }

        expect([1, 2], 'when passed as parameters to', add, 'to equal', 3);

        try {
            expect([1, 2], 'when passed as parameters to', add, 'to equal', 9);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([1, 2], 'when passed as parameters to', add, 'to equal', 9);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 1, 2 ]\n" +
                "when passed as parameters to function add(a, b) { return a + b; } to equal 9\n" +
                "  expected 3 to equal 9"
            );
        }

        testPromises.push(expect.promise(function () {
            function delayedAdd(a, b, cb) {
                setTimeout(function () {
                    cb(null, a + b);
                }, 1);
            }

            return expect([1, 2], 'when passed as parameters to async', delayedAdd, 'to equal', 3);
        }));

        function Foo(value) {
            this.value = value;
        }

        expect(123, 'when passed as parameter to constructor', Foo, 'to be a', Foo);
        return expect.promise.all(testPromises);
    });

    it("assertions/boolean/to-be-false.md contains correct examples", function () {
        var testPromises = [];
        expect(false, 'to be false');

        try {
            expect(true, 'to be false');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(true, 'to be false');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected true to be false"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/boolean/to-be-true.md contains correct examples", function () {
        var testPromises = [];
        expect(true, 'to be true');

        try {
            expect(false, 'to be true');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(false, 'to be true');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected false to be true"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-call-the-callback-with-error.md contains correct examples", function () {
        var testPromises = [];
        function myFailingAsyncFunction(cb) {
            setTimeout(function () {
                cb(new Error('Oh dear'));
            }, 0);
        }

        testPromises.push(expect.promise(function () {
            return expect(myFailingAsyncFunction, 'to call the callback with error');
        }));

        testPromises.push(expect.promise(function () {
            return expect(myFailingAsyncFunction, 'to call the callback with error', 'Oh dear');
        }));

        testPromises.push(expect.promise(function () {
            return expect(myFailingAsyncFunction, 'to call the callback with error', /dear/);
        }));

        testPromises.push(expect.promise(function () {
            return expect(myFailingAsyncFunction, 'to call the callback with error', new Error('foo'));
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("return expect(myFailingAsyncFunction, 'to call the callback with error', new Error('foo'));").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected\n" +
                "function myFailingAsyncFunction(cb) {\n" +
                "  setTimeout(function () {\n" +
                "    cb(new Error('Oh dear'));\n" +
                "  }, 0);\n" +
                "}\n" +
                "to call the callback with error Error('foo')\n" +
                "  expected Error('Oh dear') to satisfy Error('foo')\n" +
                "\n" +
                "  Error({\n" +
                "    message: 'Oh dear' // should equal 'foo'\n" +
                "                       // -Oh dear\n" +
                "                       // +foo\n" +
                "  })"
            );
        }));


        function asyncFn(cb) {
            cb(new Error('yikes'));
        }

        testPromises.push(expect.promise(function () {
            return expect(asyncFn, 'to call the callback with error').then(function (err) {
                // err will be new Error('yikes')
            });
        }));
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-call-the-callback-without-error.md contains correct examples", function () {
        var testPromises = [];
        testPromises.push(expect.promise(function () {
            function mySuccessfulAsyncFunction(cb) {
                setTimeout(cb, 0);
            }

            return expect(mySuccessfulAsyncFunction, 'to call the callback without error');
        }));

        testPromises.push(expect.promise(function () {
            function myFailingAsyncFunction(cb) {
                setTimeout(function () {
                    cb(new Error('Oh dear'));
                }, 0);
            }

            return expect(myFailingAsyncFunction, 'to call the callback without error');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("function myFailingAsyncFunction(cb) {").nl();
                    output.code("    setTimeout(function () {").nl();
                    output.code("        cb(new Error('Oh dear'));").nl();
                    output.code("    }, 0);").nl();
                    output.code("}").nl();
                    output.code("").nl();
                    output.code("return expect(myFailingAsyncFunction, 'to call the callback without error');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "expected\n" +
                "function myFailingAsyncFunction(cb) {\n" +
                "  setTimeout(function () {\n" +
                "    cb(new Error('Oh dear'));\n" +
                "  }, 0);\n" +
                "}\n" +
                "to call the callback without error\n" +
                "  called the callback with: Error('Oh dear')"
            );
        }));


        function asyncFn(cb) {
            cb(null, 123, 456);
        }

        testPromises.push(expect.promise(function () {
            return expect(asyncFn, 'to call the callback without error').then(function (args) {
                // args will be [123, 456];
            });
        }));

        testPromises.push(expect.promise(function () {
            return expect(asyncFn, 'to call the callback without error').spread(function (result1, result2) {
                // result1 will be 123
                // result2 will be 456
            });
        }));
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-call-the-callback.md contains correct examples", function () {
        var testPromises = [];
        testPromises.push(expect.promise(function () {
            function mySuccessfulAsyncFunction(cb) {
                setTimeout(function () {
                    cb();
                });
            }

            return expect(mySuccessfulAsyncFunction, 'to call the callback');
        }));

        testPromises.push(expect.promise(function () {
            function errorOut(cb) {
                throw new Error('ugh');
            }
            return expect(errorOut, 'to call the callback');
        }).then(function () {
            return expect.promise(function () {
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("function errorOut(cb) {").nl();
                    output.code("    throw new Error('ugh');").nl();
                    output.code("}").nl();
                    output.code("return expect(errorOut, 'to call the callback');").nl();
                    output.error("to throw");
                });
            });
        }).caught(function (e) {
            expect(e, "to have message",
                "ugh"
            );
        }));


        function asyncFn(cb) {
            setTimeout(function () {
                cb(null, 'foo');
            });
        }

        testPromises.push(expect.promise(function () {
            return expect(asyncFn, 'to call the callback').then(function (args) {
                // args will be [null, 'foo'];
            });
        }));

        testPromises.push(expect.promise(function () {
            return expect(asyncFn, 'to call the callback').spread(function (err, result) {
                // err will be null
                // result will be 'foo'
            });
        }));
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-error.md contains correct examples", function () {
        var testPromises = [];
        function willBeRejected() {
            return expect.promise(function (resolve, reject) {
                reject(new Error('The reject message'));
            });
        }
        function willThrow() {
            throw new Error('The error message');
        }
        expect(willBeRejected, 'to error');
        expect(willThrow, 'to error');

        try {
            function willNotBeRejected() {
                return expect.promise(function (resolve, reject) {
                    resolve('Hello world');
                });
            }
            expect(willNotBeRejected, 'to error');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("function willNotBeRejected() {").nl();
                output.code("    return expect.promise(function (resolve, reject) {").nl();
                output.code("        resolve('Hello world');").nl();
                output.code("    });").nl();
                output.code("}").nl();
                output.code("expect(willNotBeRejected, 'to error');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "function willNotBeRejected() {\n" +
                "  return expect.promise(function (resolve, reject) {\n" +
                "    resolve('Hello world');\n" +
                "  });\n" +
                "}\n" +
                "to error"
            );
        }

        expect(willBeRejected, 'to error', 'The reject message');

        try {
            expect(willBeRejected, 'to error', 'The error message');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(willBeRejected, 'to error', 'The error message');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "function willBeRejected() {\n" +
                "  return expect.promise(function (resolve, reject) {\n" +
                "    reject(new Error('The reject message'));\n" +
                "  });\n" +
                "}\n" +
                "to error 'The error message'\n" +
                "  expected Error('The reject message') to satisfy 'The error message'\n" +
                "\n" +
                "  -The reject message\n" +
                "  +The error message"
            );
        }

        expect(willBeRejected, 'to error', /reject message/);

        try {
            expect(willBeRejected, 'to error', /error message/);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(willBeRejected, 'to error', /error message/);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "function willBeRejected() {\n" +
                "  return expect.promise(function (resolve, reject) {\n" +
                "    reject(new Error('The reject message'));\n" +
                "  });\n" +
                "}\n" +
                "to error /error message/\n" +
                "  expected Error('The reject message') to satisfy /error message/"
            );
        }

        expect(willNotBeRejected, 'not to error');

        try {
            expect(willBeRejected, 'not to error');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(willBeRejected, 'not to error');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected\n" +
                "function willBeRejected() {\n" +
                "  return expect.promise(function (resolve, reject) {\n" +
                "    reject(new Error('The reject message'));\n" +
                "  });\n" +
                "}\n" +
                "not to error\n" +
                "  returned promise rejected with: Error('The reject message')"
            );
        }

        testPromises.push(expect.promise(function () {
            function willBeRejectedAsync() {
                return expect.promise(function (resolve, reject) {
                    setTimeout(function () {
                        reject(new Error('async error'));
                    }, 1);
                });
            }

            return expect(willBeRejectedAsync, 'to error', function (e) {
                return expect(e.message, 'to equal', 'async error');
            });
        }));

        testPromises.push(expect.promise(function () {
            var errorCount = 0;
            function willBeRejectedAsync() {
                return expect.promise(function (resolve, reject) {
                    setTimeout(function () {
                        var error = new Error('async error');
                        errorCount += 1;
                        error.errorCount = errorCount;
                        reject(error);
                    }, 1);
                });
            }

            return expect(willBeRejectedAsync, 'to error', function (e) {
                return expect(willBeRejectedAsync, 'to error', function (e2) {
                    return expect(e2.errorCount, 'to be greater than', e.errorCount);
                });
            });
        }));
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-have-arity.md contains correct examples", function () {
        var testPromises = [];
        expect(Math.max, 'to have arity', 2);
        expect('wat'.substring, 'to have arity', 2);

        try {
            expect(function wat(foo, bar) {
              return foo + bar;
            }, 'to have arity', 3);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(function wat(foo, bar) {").nl();
                output.code("  return foo + bar;").nl();
                output.code("}, 'to have arity', 3);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected function wat(foo, bar) { return foo + bar; } to have arity 3"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/function/to-throw.md contains correct examples", function () {
        var testPromises = [];
        function willThrow() {
          throw new Error('The error message');
        }
        expect(willThrow, 'to throw');
        expect(willThrow, 'to throw error');
        expect(willThrow, 'to throw exception');

        try {
            expect(function willNotThrow() {}, 'to throw');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(function willNotThrow() {}, 'to throw');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected function willNotThrow() {} to throw"
            );
        }

        expect(function () {
          throw new Error('The error message');
        }, 'to throw', 'The error message');

        if (!isPhantom) {
            try {
                expect(function () {
                  throw new Error('The error message!');
                }, 'to throw', 'The error message');
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(function () {").nl();
                    output.code("  throw new Error('The error message!');").nl();
                    output.code("}, 'to throw', 'The error message');").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected\n" +
                    "function () {\n" +
                    "  throw new Error('The error message!');\n" +
                    "}\n" +
                    "to throw 'The error message'\n" +
                    "  expected Error('The error message!') to satisfy 'The error message'\n" +
                    "\n" +
                    "  -The error message!\n" +
                    "  +The error message"
                );
            }
        }

        expect(function () {
          throw new Error('The error message');
        }, 'to throw', /error message/);

        if (!isPhantom) {
            try {
                expect(function () {
                  throw new Error('The error message!');
                }, 'to throw', /catastrophic failure/);
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(function () {").nl();
                    output.code("  throw new Error('The error message!');").nl();
                    output.code("}, 'to throw', /catastrophic failure/);").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected\n" +
                    "function () {\n" +
                    "  throw new Error('The error message!');\n" +
                    "}\n" +
                    "to throw /catastrophic failure/\n" +
                    "  expected Error('The error message!') to satisfy /catastrophic failure/"
                );
            }
        }

        expect(function () {
          this.foo.bar();
        }, 'to throw', function (e) {
          expect(e, 'to be a', TypeError);
        });

        if (!isPhantom) {
            try {
                expect(function () {
                  throw new Error('Another error');
                }, 'to throw', function (e) {
                  expect(e, 'to be a', TypeError);
                });
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(function () {").nl();
                    output.code("  throw new Error('Another error');").nl();
                    output.code("}, 'to throw', function (e) {").nl();
                    output.code("  expect(e, 'to be a', TypeError);").nl();
                    output.code("});").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected\n" +
                    "function () {\n" +
                    "  throw new Error('Another error');\n" +
                    "}\n" +
                    "to throw\n" +
                    "function (e) {\n" +
                    "  expect(e, 'to be a', TypeError);\n" +
                    "}\n" +
                    "  expected Error('Another error') to be a TypeError"
                );
            }
        }

        expect(function () {
          throw new TypeError('Invalid syntax');
        }, 'to throw', new TypeError('Invalid syntax'));

        if (!isPhantom) {
            try {
                expect(function () {
                  throw new Error('Another error');
                }, 'to throw', new TypeError('Invalid syntax'));
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(function () {").nl();
                    output.code("  throw new Error('Another error');").nl();
                    output.code("}, 'to throw', new TypeError('Invalid syntax'));").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected\n" +
                    "function () {\n" +
                    "  throw new Error('Another error');\n" +
                    "}\n" +
                    "to throw TypeError('Invalid syntax')\n" +
                    "  expected Error('Another error') to satisfy TypeError('Invalid syntax')"
                );
            }
        }

        expect(function () {
          // Do some work that should not throw
        }, 'not to throw');

        if (!isPhantom) {
            try {
                expect(function () {
                  throw new Error('threw anyway');
                }, 'not to throw');
                expect.fail(function (output) {
                    output.error("expected:").nl();
                    output.code("expect(function () {").nl();
                    output.code("  throw new Error('threw anyway');").nl();
                    output.code("}, 'not to throw');").nl();
                    output.error("to throw");
                });
            } catch (e) {
                expect(e, "to have message",
                    "expected\n" +
                    "function () {\n" +
                    "  throw new Error('threw anyway');\n" +
                    "}\n" +
                    "not to throw\n" +
                    "  threw: Error('threw anyway')"
                );
            }
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-NaN.md contains correct examples", function () {
        var testPromises = [];
        expect(NaN, 'to be NaN');

        try {
            expect(2, 'to be NaN');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'to be NaN');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 to be NaN"
            );
        }

        expect(2, 'not to be NaN');

        try {
            expect(NaN, 'not to be NaN');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(NaN, 'not to be NaN');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected NaN not to be NaN"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-close-to.md contains correct examples", function () {
        var testPromises = [];
        expect(1.5, 'to be close to', 1.500001, 1e-5);
        expect(1.5, 'to be close to', 1.5000000001)

        try {
            expect(1.5, 'to be close to', 1.50001, 1e-5);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1.5, 'to be close to', 1.50001, 1e-5);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1.5 to be close to 1.50001 (epsilon: 1e-5)"
            );
        }

        expect(1.5, 'not to be close to', 1.499, 1e-4);

        try {
            expect(1.5, 'not to be close to', 1.5000000001)
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1.5, 'not to be close to', 1.5000000001)").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1.5 not to be close to 1.5000000001 (epsilon: 1e-9)"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-finite.md contains correct examples", function () {
        var testPromises = [];
        expect(123, 'to be finite');

        try {
            expect(Infinity, 'to be finite');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(Infinity, 'to be finite');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected Infinity to be finite"
            );
        }

        expect(Infinity, 'not to be finite');
        expect(-Infinity, 'not to be finite');

        try {
            expect(123, 'not to be finite');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(123, 'not to be finite');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 123 not to be finite"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-greater-than-or-equal-to.md contains correct examples", function () {
        var testPromises = [];
        expect(3, 'to be greater than or equal to', 3);

        try {
            expect(1, 'to be greater than or equal to', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1, 'to be greater than or equal to', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1 to be greater than or equal to 2"
            );
        }

        expect(1, 'not to be greater than or equal to', 2);

        try {
            expect(2, 'not to be greater than or equal to', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'not to be greater than or equal to', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 not to be greater than or equal to 2"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-greater-than.md contains correct examples", function () {
        var testPromises = [];
        expect(3, 'to be greater than', 2);
        expect(1, 'to be above', 0);

        try {
            expect(2, 'to be greater than', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'to be greater than', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 to be greater than 2"
            );
        }

        expect(2, 'not to be greater than', 2);
        expect(0, 'not to be above', 1);

        try {
            expect(3, 'not to be greater than', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(3, 'not to be greater than', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 3 not to be greater than 2"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-infinite.md contains correct examples", function () {
        var testPromises = [];
        expect(Infinity, 'to be infinite');
        expect(-Infinity, 'to be infinite');

        try {
            expect(123, 'to be infinite');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(123, 'to be infinite');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 123 to be infinite"
            );
        }

        expect(123, 'not to be infinite');

        try {
            expect(Infinity, 'not to be infinite');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(Infinity, 'not to be infinite');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected Infinity not to be infinite"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-less-than-or-equal-to.md contains correct examples", function () {
        var testPromises = [];
        expect(3, 'to be less than or equal to', 3);

        try {
            expect(2, 'to be less than or equal to', 1);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'to be less than or equal to', 1);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 to be less than or equal to 1"
            );
        }

        expect(2, 'not to be less than or equal to', 1);

        try {
            expect(2, 'not to be less than or equal to', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'not to be less than or equal to', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 not to be less than or equal to 2"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-less-than.md contains correct examples", function () {
        var testPromises = [];
        expect(2, 'to be less than', 3);
        expect(0, 'to be below', 1);

        try {
            expect(2, 'to be less than', 2);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'to be less than', 2);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 to be less than 2"
            );
        }

        expect(2, 'not to be less than', 2);
        expect(1, 'not to be below', 0);

        try {
            expect(2, 'not to be less than', 3);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(2, 'not to be less than', 3);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 2 not to be less than 3"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-negative.md contains correct examples", function () {
        var testPromises = [];
        expect(-42, 'to be negative');

        try {
            expect(0, 'to be negative');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(0, 'to be negative');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 0 to be negative"
            );
        }

        expect(0, 'not to be negative');
        expect(42, 'not to be negative');

        try {
            expect(-1, 'not to be negative');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(-1, 'not to be negative');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected -1 not to be negative"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-positive.md contains correct examples", function () {
        var testPromises = [];
        expect(42, 'to be positive');

        try {
            expect(0, 'to be positive');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(0, 'to be positive');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 0 to be positive"
            );
        }

        expect(0, 'not to be positive');
        expect(-42, 'not to be positive');

        try {
            expect(1, 'not to be positive');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(1, 'not to be positive');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 1 not to be positive"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/number/to-be-within.md contains correct examples", function () {
        var testPromises = [];
        expect(0, 'to be within', 0, 4);
        expect(1, 'to be within', 0, 4);
        expect(2.5, 'to be within', 0, 4);
        expect(4, 'to be within', 0, 4);

        try {
            expect(-1, 'to be within', 0, 4);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(-1, 'to be within', 0, 4);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected -1 to be within 0..4"
            );
        }

        expect(-1, 'not to be within', 0, 4);
        expect(5, 'not to be within', 0, 4);

        try {
            expect(0, 'not to be within', 0, 4);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(0, 'not to be within', 0, 4);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 0 not to be within 0..4"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-be-canonical.md contains correct examples", function () {
        var testPromises = [];
        expect({ a: 123, b: 456 }, 'to be canonical');
        expect([456, { a: 123 }], 'to be canonical');

        try {
            expect([456, { b: 456, a: 123 }], 'to be canonical');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect([456, { b: 456, a: 123 }], 'to be canonical');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected [ 456, { b: 456, a: 123 } ] to be canonical"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-key.md contains correct examples", function () {
        var testPromises = [];
        expect({ a: 'a', b: 'b', c: 'c' }, 'to have key', 'a');

        expect({ a: 'a' }, 'to only have key', 'a');

        try {
            expect({ a: 'a', b: 'b' }, 'to have key', 'c');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'a', b: 'b' }, 'to have key', 'c');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'a', b: 'b' } to have key 'c'"
            );
        }

        expect({ a: 'a', b: 'b' }, 'not to have key', 'c');
        expect(Object.create({ a: 'a', b: 'b' }), 'not to have key', 'a');

        expect({ a: 'a', b: 'b' }, 'to not only have key', 'a');

        try {
            expect({ a: 'a', b: 'b' }, 'to not have key', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'a', b: 'b' }, 'to not have key', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'a', b: 'b' } to not have key 'a'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-keys-satisfying.md contains correct examples", function () {
        var testPromises = [];
        expect({ foo: 0, bar: 1, baz: 2, qux: 3 },
               'to have keys satisfying', function (key) {
            expect(key, 'to match', /^[a-z]{3}$/);
        });

        expect({ foo: 0, bar: 1, baz: 2, qux: 3 },
               'to have keys satisfying',
               'to match', /^[a-z]{3}$/);

        try {
            expect({ foo: 0, bar: 1, baz: 2, qux: 3, quux: 4 },
                   'to have keys satisfying',
                   'to match', /^[a-z]{3}$/);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ foo: 0, bar: 1, baz: 2, qux: 3, quux: 4 },").nl();
                output.code("       'to have keys satisfying',").nl();
                output.code("       'to match', /^[a-z]{3}$/);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { foo: 0, bar: 1, baz: 2, qux: 3, quux: 4 }\n" +
                "to have keys satisfying 'to match', /^[a-z]{3}$/\n" +
                "\n" +
                "[\n" +
                "  'foo',\n" +
                "  'bar',\n" +
                "  'baz',\n" +
                "  'qux',\n" +
                "  'quux' // should match /^[a-z]{3}$/\n" +
                "]"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-keys.md contains correct examples", function () {
        var testPromises = [];
        expect({ a: 'a', b: 'b', c: 'c' }, 'to have keys', 'a', 'c');
        expect({ a: 'a', b: 'b', c: 'c' }, 'to have keys', ['a', 'c']);

        expect({ a: 'a', b: 'b', c: 'c' }, 'to only have keys', ['a', 'c', 'b']);

        try {
            expect({ a: 'a', b: 'b', c: 'c' }, 'to have keys', 'c', 'd');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'a', b: 'b', c: 'c' }, 'to have keys', 'c', 'd');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'a', b: 'b', c: 'c' } to have keys 'c', 'd'"
            );
        }

        expect({ a: 'a', b: 'b', c: 'c' }, 'not to have keys', 'd', 'e');
        expect(Object.create({ a: 'a', b: 'b', c: 'c' }), 'not to have keys', 'a', 'b');

        expect({ a: 'a', b: 'b', c: 'c' }, 'to not only have keys', 'a', 'b');

        try {
            expect({ a: 'a', b: 'b', c: 'c' }, 'to not only have keys', 'a', 'b', 'c');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'a', b: 'b', c: 'c' }, 'to not only have keys', 'a', 'b', 'c');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'a', b: 'b', c: 'c' } to not only have keys 'a', 'b', 'c'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-properties.md contains correct examples", function () {
        var testPromises = [];
        expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'to have properties', ['a', 'b']);
        expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'to have properties', {
            a: 'a',
            b: { c: 'c' }
        });
        expect([ 'a', { c: 'c' }, 'd' ], 'to have properties', {
            1: { c: 'c' },
            2: 'd'
        });

        expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'to have own properties', ['a', 'b']);
        expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'to have own properties', {
            a: 'a',
            b: { c: 'c' }
        });

        try {
            expect({ a: 'f00', b: 'bar' }, 'to have own properties', {
              a: 'foo',
              c: 'baz'
            });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'f00', b: 'bar' }, 'to have own properties', {").nl();
                output.code("  a: 'foo',").nl();
                output.code("  c: 'baz'").nl();
                output.code("});").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'f00', b: 'bar' } to have own properties { a: 'foo', c: 'baz' }\n" +
                "\n" +
                "{\n" +
                "  a: 'f00', // should equal 'foo'\n" +
                "            // -f00\n" +
                "            // +foo\n" +
                "  b: 'bar',\n" +
                "  c: undefined // should equal 'baz'\n" +
                "}"
            );
        }

        expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'not to have properties', ['k', 'l']);
        expect(Object.create({ a: 'a', b: 'b' }), 'not to have own properties', ['a', 'b']);

        try {
            expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'not to have properties', ['b', 'd']);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ a: 'a', b: { c: 'c' }, d: 'd' }, 'not to have properties', ['b', 'd']);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { a: 'a', b: { c: 'c' }, d: 'd' } not to have properties [ 'b', 'd' ]"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-property.md contains correct examples", function () {
        var testPromises = [];
        expect([1, 2], 'to have property', 'length');
        expect({ a: 'b' }, 'to have property', 'a');
        expect({ a: 'b' }, 'to have property', 'toString');

        expect([1, 2], 'to have property', 'length', 2);
        expect({ a: 'b' }, 'to have property', 'a', 'b');
        expect({ a: { b: 'c' } }, 'to have property', 'a', { b: 'c' });

        expect({ a: 'b' }, 'to have own property', 'a');
        expect({ a: 'b' }, 'to have own property', 'a', 'b');

        try {
            expect(Object.create({ a: 'b' }), 'to have own property', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect(Object.create({ a: 'b' }), 'to have own property', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected {} to have own property 'a'"
            );
        }

        expect({ a: 'b' }, 'not to have property', 'b');
        expect(Object.create({ a: 'b' }), 'not to have own property', 'a');
        return expect.promise.all(testPromises);
    });

    it("assertions/object/to-have-values-satisfying.md contains correct examples", function () {
        var testPromises = [];
        expect({ foo: 0, bar: 1, baz: 2, qux: 3 },
               'to have values satisfying', function (value, index) {
            expect(value, 'to be a number');
        });

        expect({ foo: 0, bar: 1, baz: 2, qux: 3 },
               'to have values satisfying',
               'to be a number');

        try {
            expect({ foo: [0, 1, 2], bar: [4, 5, 6], baz: [7, 8, 9] },
                   'to have values satisfying',
                   'to have items satisfying',
                   expect.it('to be a number').and('to be below', 8));
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ foo: [0, 1, 2], bar: [4, 5, 6], baz: [7, 8, 9] },").nl();
                output.code("       'to have values satisfying',").nl();
                output.code("       'to have items satisfying',").nl();
                output.code("       expect.it('to be a number').and('to be below', 8));").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected object to have values satisfying\n" +
                "'to have items satisfying', expect.it('to be a number')\n" +
                "        .and('to be below', 8)\n" +
                "\n" +
                "{\n" +
                "  foo: [ 0, 1, 2 ],\n" +
                "  bar: [ 4, 5, 6 ],\n" +
                "  baz: [\n" +
                "    7,\n" +
                "    8, // ✓ should be a number and\n" +
                "       // ⨯ should be below 8\n" +
                "    9 // ✓ should be a number and\n" +
                "      // ⨯ should be below 8\n" +
                "  ]\n" +
                "}"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-empty.md contains correct examples", function () {
        var testPromises = [];
        expect('', 'to be empty');

        try {
            expect('Hello world', 'to be empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world', 'to be empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world' to be empty"
            );
        }

        expect('Hello world', 'not to be empty');

        try {
            expect('', 'not to be empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('', 'not to be empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '' not to be empty"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-greater-than-or-equal-to.md contains correct examples", function () {
        var testPromises = [];
        expect('b', 'to be greater than or equal to', 'b');

        try {
            expect('a', 'to be greater than or equal to', 'b');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'to be greater than or equal to', 'b');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' to be greater than or equal to 'b'"
            );
        }

        expect('a', 'not to be greater than or equal to', 'b');

        try {
            expect('a', 'not to be greater than or equal to', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'not to be greater than or equal to', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' not to be greater than or equal to 'a'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-greater-than.md contains correct examples", function () {
        var testPromises = [];
        expect('b', 'to be greater than', 'a');
        expect('b', 'to be above', 'a');

        try {
            expect('a', 'to be greater than', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'to be greater than', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' to be greater than 'a'"
            );
        }

        expect('a', 'not to be greater than', 'a');
        expect('a', 'not to be above', 'a');

        try {
            expect('b', 'not to be above', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('b', 'not to be above', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'b' not to be above 'a'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-less-than-or-equal-to.md contains correct examples", function () {
        var testPromises = [];
        expect('b', 'to be less than or equal to', 'b');

        try {
            expect('b', 'to be less than or equal to', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('b', 'to be less than or equal to', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'b' to be less than or equal to 'a'"
            );
        }

        expect('b', 'not to be less than or equal to', 'a');

        try {
            expect('a', 'not to be less than or equal to', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'not to be less than or equal to', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' not to be less than or equal to 'a'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-less-than.md contains correct examples", function () {
        var testPromises = [];
        expect('a', 'to be less than', 'b');
        expect('a', 'to be below', 'b');

        try {
            expect('a', 'to be less than', 'a');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'to be less than', 'a');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' to be less than 'a'"
            );
        }

        expect('a', 'not to be less than', 'a');
        expect('a', 'not to be below', 'a');

        try {
            expect('a', 'not to be below', 'b');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('a', 'not to be below', 'b');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'a' not to be below 'b'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-non-empty.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello', 'to be non-empty');

        try {
            expect('', 'to be non-empty');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('', 'to be non-empty');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected '' to be non-empty"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be-within.md contains correct examples", function () {
        var testPromises = [];
        expect('a', 'to be within', 'a', 'd');
        expect('b', 'to be within', 'a', 'd');
        expect('aabbcc', 'to be within', 'aaa', 'aaz');

        try {
            expect('abbbcc', 'to be within', 'aaa', 'aaz');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('abbbcc', 'to be within', 'aaa', 'aaz');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'abbbcc' to be within 'aaa'..'aaz'"
            );
        }

        expect('bar', 'not to be within', 'foo', 'baz');
        expect('e', 'not to be within', 'a', 'd');
        expect('abbbcc', 'not to be within', 'aaa', 'aaz');

        try {
            expect('c', 'not to be within', 'a', 'd');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('c', 'not to be within', 'a', 'd');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'c' not to be within 'a'..'d'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-be.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello', 'to be', 'Hello');

        try {
            expect('Hello beautiful!', 'to be', 'Hello world!');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello beautiful!', 'to be', 'Hello world!');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello beautiful!' to be 'Hello world!'\n" +
                "\n" +
                "-Hello beautiful!\n" +
                "+Hello world!"
            );
        }

        expect('Hello', 'not to be', 'Hello world!');
        expect('1', 'not to be', 1);

        try {
            expect('Hello world!', 'not to be', 'Hello world!');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'not to be', 'Hello world!');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' not to be 'Hello world!'"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-begin-with.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello beautiful world!', 'to begin with', 'Hello');

        try {
            expect('Hello world!', 'to begin with', 'foo');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to begin with', 'foo');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to begin with 'foo'"
            );
        }

        try {
            expect('Hello world!', 'to begin with', 'Hell yeah');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to begin with', 'Hell yeah');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to begin with 'Hell yeah'\n" +
                "\n" +
                "Hello world!\n" +
                "^^^^"
            );
        }

        expect('Hello world!', 'not to begin with', 'Heaven');

        try {
            expect('Hello beautiful world!', 'not to begin with', 'Hello');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello beautiful world!', 'not to begin with', 'Hello');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello beautiful world!' not to begin with 'Hello'\n" +
                "\n" +
                "Hello beautiful world!\n" +
                "^^^^^"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-contain.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello beautiful world!', 'to contain', 'beautiful');
        expect('Hello beautiful world!', 'to contain', 'Hello', 'world');

        try {
            expect('Hello world!', 'to contain', 'beautiful');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to contain', 'beautiful');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to contain 'beautiful'\n" +
                "\n" +
                "Hello world!"
            );
        }

        try {
            expect('BEGIN:VEVENT\n' +
                   'UID:B4D3A0F8-6E38-5F0E-8CE7-BA0171DEAA8D\n' +
                   'RECURRENCE-ID;TZID=Europe/Copenhagen:20150113T120000\n' +
                   'DTSTART;TZID=Europe/Copenhagen:20150203T110000\n' +
                   'DTEND;TZID=Europe/Copenhagen:20150203T120000\n' +
                   'DTSTAMP:20150114T121235Z\n' +
                   'SUMMARY:Lunch\n' +
                   'LOCATION:Kalvebod Brygge 24, Copenhagen\n' +
                   'END:VEVENT',
                   'to contain',
                   'UID:6FA459EA-EE8A-3CA4-894E-DB77E160355E');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('BEGIN:VEVENT\\n' +").nl();
                output.code("       'UID:B4D3A0F8-6E38-5F0E-8CE7-BA0171DEAA8D\\n' +").nl();
                output.code("       'RECURRENCE-ID;TZID=Europe/Copenhagen:20150113T120000\\n' +").nl();
                output.code("       'DTSTART;TZID=Europe/Copenhagen:20150203T110000\\n' +").nl();
                output.code("       'DTEND;TZID=Europe/Copenhagen:20150203T120000\\n' +").nl();
                output.code("       'DTSTAMP:20150114T121235Z\\n' +").nl();
                output.code("       'SUMMARY:Lunch\\n' +").nl();
                output.code("       'LOCATION:Kalvebod Brygge 24, Copenhagen\\n' +").nl();
                output.code("       'END:VEVENT',").nl();
                output.code("       'to contain',").nl();
                output.code("       'UID:6FA459EA-EE8A-3CA4-894E-DB77E160355E');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'BEGIN:VEVENT\\nUID:B4D3A0F8-6E38-5F0E-8CE7-BA0171DEAA8D\\nRECURRENCE-ID;TZID=Europe/Copenhagen:20150113T120000\\nDTSTART;TZID=Europe/Copenhagen:20150203T110000\\nDTEND;TZID=Europe/Copenhagen:20150203T120000\\nDTSTAMP:20150114T121235Z\\nSUMMARY:Lunch\\nLOCATION:Kalvebod Brygge 24, Copenhagen\\nEND:VEVENT'\n" +
                "to contain 'UID:6FA459EA-EE8A-3CA4-894E-DB77E160355E'\n" +
                "\n" +
                "BEGIN:VEVENT\n" +
                "UID:B4D3A0F8-6E38-5F0E-8CE7-BA0171DEAA8D\n" +
                "^^^^\n" +
                "RECURRENCE-ID;TZID=Europe/Copenhagen:20150113T120000\n" +
                "DTSTART;TZID=Europe/Copenhagen:20150203T110000\n" +
                "DTEND;TZID=Europe/Copenhagen:20150203T120000\n" +
                "DTSTAMP:20150114T121235Z\n" +
                "SUMMARY:Lunch\n" +
                "LOCATION:Kalvebod Brygge 24, Copenhagen\n" +
                "END:VEVENT"
            );
        }

        expect('Hello world!', 'not to contain', 'beautiful', 'ugly');

        try {
            expect('Hello beautiful world!', 'not to contain', 'beautiful', 'ugly');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello beautiful world!', 'not to contain', 'beautiful', 'ugly');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello beautiful world!' not to contain 'beautiful', 'ugly'\n" +
                "\n" +
                "Hello beautiful world!\n" +
                "      ^^^^^^^^^"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-end-with.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello beautiful world!', 'to end with', 'world!');

        try {
            expect('Hello world!', 'to end with', 'foo');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to end with', 'foo');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to end with 'foo'"
            );
        }

        try {
            expect('Hello world!', 'to end with', 'Hola, world!');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to end with', 'Hola, world!');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to end with 'Hola, world!'\n" +
                "\n" +
                "Hello world!\n" +
                "     ^^^^^^^"
            );
        }

        expect('Hello world!', 'not to end with', 'earth!');

        try {
            expect('Hello beautiful world!', 'not to end with', 'world!');
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello beautiful world!', 'not to end with', 'world!');").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello beautiful world!' not to end with 'world!'\n" +
                "\n" +
                "Hello beautiful world!\n" +
                "                ^^^^^^"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-have-length.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello world', 'to have length', 11);

        try {
            expect('Hello world', 'to have length', 12);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world', 'to have length', 12);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world' to have length 12\n" +
                "  expected 11 to be 12"
            );
        }

        expect('Hello world', 'not to have length', 12);

        try {
            expect('Hello world', 'not to have length', 11);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world', 'not to have length', 11);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world' not to have length 11"
            );
        }
        return expect.promise.all(testPromises);
    });

    it("assertions/string/to-match.md contains correct examples", function () {
        var testPromises = [];
        expect('Hello beautiful world!', 'to match', /bea.t.*/);

        try {
            expect('Hello world!', 'to match', /beautiful/);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello world!', 'to match', /beautiful/);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello world!' to match /beautiful/"
            );
        }

        expect('Hello world!', 'not to match', /beautiful/);

        try {
            expect('Hello beautiful world!', 'not to match', /beautiful/);
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect('Hello beautiful world!', 'not to match', /beautiful/);").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected 'Hello beautiful world!' not to match /beautiful/\n" +
                "\n" +
                "Hello beautiful world!\n" +
                "      ^^^^^^^^^"
            );
        }

        testPromises.push(expect.promise(function () {
            return expect('Hello world!', 'to match', /(\w+)!/).then(function (captures) {
                expect(captures[0], 'to equal', 'world!');
                expect(captures[1], 'to equal', 'world');
                expect(captures.index, 'to equal', 6);
            });
        }));

        testPromises.push(expect.promise(function () {
            return expect('Hello world!', 'to match', /(\w+)!/).spread(function ($0, $1) {
                expect($0, 'to equal', 'world!');
                expect($1, 'to equal', 'world');
            });
        }));
        return expect.promise.all(testPromises);
    });

    it("index.md contains correct examples", function () {
        var testPromises = [];
        try {
            expect({ text: 'f00!' }, 'to equal', { text: 'foo!' });
            expect.fail(function (output) {
                output.error("expected:").nl();
                output.code("expect({ text: 'f00!' }, 'to equal', { text: 'foo!' });").nl();
                output.error("to throw");
            });
        } catch (e) {
            expect(e, "to have message",
                "expected { text: 'f00!' } to equal { text: 'foo!' }\n" +
                "\n" +
                "{\n" +
                "  text: 'f00!' // should equal 'foo!'\n" +
                "               // -f00!\n" +
                "               // +foo!\n" +
                "}"
            );
        }
        return expect.promise.all(testPromises);
    });
});
